### PR TITLE
imagebuilder: split Executor and per-stage execution

### PR DESCRIPTION
**Pull request by nalind**
When building images using multiple stages, we were using a shallow copy of the `Executor`, depending on some fields to be shared between all instances, while replacing others fields with per-stage values.

This patch makes that easier for me to reason about by introducing a `StageExecutor` which holds the stage-specific information, and which can hold a reference to the main `Executor`, which now coordinates stages using `StageExecutors` rather than `Executors`.

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
@vrothberg @giuseppe @TomSweeneyRedHat @umohnani8 PTAL

**Comment by umohnani8** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
build issue appears to be a timeout flake.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please

**Comment by TomSweeneyRedHat** [comment]
passes baseline and the other tests I threw at it.  LGTM, but would like a head nod from @vrothberg in the morning.

**Comment by vrothberg** [comment]
I restarted the two failed tests, so we can merge the PR once everything's green.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by nalind** [comment]
@rh-atomic-bot retry

**Comment by nalind** [comment]
Rebased.

**Comment by nalind** [comment]
Removed logic to reset the stage's volume cache each time we commit a layer in a stage - the cached contents of parts of the working container's rootfs don't need to be regenerated just because we committed a layer and created a new container from the image, since we know that that portion of the new container;s directory tree is supposed to have the same contents as the old one.

File: imagebuildah/build.go [hunk 1]
```diff
@@ -222,9 +214,18 @@ type Executor struct {
...
+
+	index           int
+	name            string
+	builder         *buildah.Builder
+	preserved       int
+	volumes         imagebuilder.VolumeSet
+	volumeCache     map[string]string
+	volumeCacheInfo map[string]os.FileInfo
+	mountPoint      string
+	copyFrom        string // Used to keep track of the --from flag from COPY and ADD
```
Comment (rhatdan) [inline imagebuildah/build.go]: Why are you doing this?

Comment (nalind) [reply imagebuildah/build.go]: We were making shallow copies of this structure in `withName()` and then depending on some of the fields to remain shared between them, and some to not be.  It was getting difficult to debug.

Comment (rhatdan) [reply imagebuildah/build.go]: Can you add a comment about that in the place of the break to explain it, so someone like me does not come back in and merge them back together.

Comment (nalind) [reply imagebuildah/build.go]: Expanded on the comment there.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**